### ISSUE TITLE
mask.py

### ISSUE THREAD

**Issue by HoangHoang1408** [HoangHoang1408/temp#28]
from dataclasses import dataclass
import math
import sys
from typing import Literal
import torch
from abc import ABC, abstractmethod
from torch import nn
import torch.nn.functional as F

class MaskBase(nn.Module, ABC):
    @property
    @abstractmethod
    def n(self):
        pass
    
    @abstractmethod
    def compute_constraint(self, target):
        pass

class TernaryMask(MaskBase, nn.Module):
    def __init__(self, *shape):
        super().__init__()

        self.mask_latent = nn.Parameter(torch.ones(*shape))
        self.lagmul_1 = nn.Parameter(torch.zeros([]))
        self.lagmul_2 = nn.Parameter(torch.zeros([]))

    def compute_constraint(self, target):
        lagmul_1 = 2 * self.lagmul_1.detach() - self.lagmul_1
        lagmul_2 = 2 * self.lagmul_2.detach() - self.lagmul_2
        c = (self.n - target) / self.m
        
        # Dual restarts
        if c.abs() <= 1e-12:
            self.lagmul_1.data = torch.zeros_like(self.lagmul_1)
            self.lagmul_2.data = torch.zeros_like(self.lagmul_2)
            
        return lagmul_1 * c + lagmul_2 * c.pow(2)

    @property
    def ternary_mask(self):
        m = self.mask_latent
        
        absmean = m.abs().mean()
        eps = 1e-5
        
        m_norm = m / (absmean + eps)
        m_ternarized = m_norm.round().clamp(-1, 1).detach()
        
        return m_ternarized + m - m.detach()

    @property
    def n(self):
        mask = self.ternary_mask
        return mask.abs().sum()

    @property
    def m(self):
        return self.mask_latent.numel()

    def forward(self, input):
        mask = self.ternary_mask
        return input * mask

class ShearingMask(MaskBase, nn.Module):
    def __init__(self, *shape: list) -> None:
        super().__init__()

        self.limit_a, self.limit_b, self.epsilon = -0.1, 1.1, 1e-6

        self.droprate_init = 0.5
        self.temperature = 2. / 3
        self.magical_number = 0.8

        self.z_loga = self.initialize_mask(shape)
        self.mask_size = self.z_loga.numel()

        self.lagmul_1 = nn.Parameter(torch.zeros([]))
        self.lagmul_2 = nn.Parameter(torch.zeros([]))

    def param_init_fn(self, module):
        """ Initialize the parameters for masking variables. """
        mean = math.log(1 - self.droprate_init) - math.log(self.droprate_init)
        mean = 5
        if isinstance(module, nn.Parameter):
            module.data.normal_(mean, 1e-2)
        else:
            for tensor in module.parameters():
                tensor.data.normal_(mean, 1e-2)

    def initialize_mask(self, mask_shape: list):
        """ Initialize the parameters for masking variables. """
        z_loga = nn.Parameter(torch.ones(*mask_shape))
        self.param_init_fn(z_loga)
        return z_loga

    def cdf_qz(self, z_loga: torch.Tensor = None):
        """ Implements the CDF of the 'stretched' concrete distribution """
        if z_loga is None:
            z_loga = self.z_loga
        xn = (0 - self.limit_a) / (self.limit_b - self.limit_a)
        logits = math.log(xn) - math.log(1 - xn)
        return torch.sigmoid(logits * self.temperature - z_loga).clamp(min=self.epsilon, max=1 - self.epsilon)

    def get_eps(self, size: list):
        """Uniform random numbers for the concrete distribution """
        eps = torch.FloatTensor(size).uniform_(self.epsilon, 1-self.epsilon)
        return eps

    def quantile_concrete(self, eps: torch.Tensor):
        y = torch.sigmoid(
            (torch.log(eps) - torch.log(1 - eps) + self.z_loga) / self.temperature)
        return y * (self.limit_b - self.limit_a) + self.limit_a

    def sample_z(self):
        eps = self.get_eps(torch.FloatTensor(
            *self.z_loga.shape)).to(self.z_loga.device)
        z = self.quantile_concrete(eps)
        z = F.hardtanh(z, min_val=0, max_val=1)
        return z

    def _deterministic_z(self, z_loga):
        expected_score = 1 - self.cdf_qz(z_loga)
        expected_num_nonzeros = expected_score.sum()
        expected_num_zeros = z_loga.nelement() - expected_num_nonzeros.item()

        try:
            num_zeros = round(expected_num_zeros)
        except:
            print("#zeros is nan....")
            sys.exit()
        soft_mask = torch.sigmoid(
            z_loga / self.temperature * self.magical_number)
        if num_zeros > 0:
            if soft_mask.ndim == 0:
                soft_mask = torch.tensor(0).to(self.z_loga.device)
            else:
                _, indices = torch.topk(soft_mask, k=num_zeros, largest=False)
                soft_mask[indices] = 0.
        return soft_mask

    def deterministic_z(self):
        if self.z_loga.ndim == 1:
            z = self._deterministic_z(self.z_loga)
        else:
            z_loga = self.z_loga.reshape(-1, self.z_loga.shape[-1])
            z = []
            for i in range(z_loga.shape[0]):
                z_ = self._deterministic_z(z_loga[i])
                z.append(z_)
            z = torch.stack(z)
        return z
    
    @property
    def n(self):
        return self.z_loga.exp().sum()

    def forward(self, input):
        if self.training:
            z = self.sample_z()
        else:
            z = self.deterministic_z()
            
        return z * input

    def constrain_parameters(self):
        self.z_loga.data.clamp_(min=math.log(1e-2), max=math.log(1e2))

    def calculate_expected_score_sparsity(self):
        score = 1 - self.cdf_qz()
        sparsity = 1 - score.sum() / self.mask_size
        return score, sparsity

    def compute_constraint(self, target):
        lagmul_1 = 2 * self.lagmul_1.detach() - self.lagmul_1
        lagmul_2 = 2 * self.lagmul_2.detach() - self.lagmul_2
        
        _, expected_sparsity = self.calculate_expected_score_sparsity()
        target_sparsity = 1 - target / self.mask_size
        c = expected_sparsity - target_sparsity
        
        return lagmul_1 * c + lagmul_2 * c.pow(2)

MASK_CLASS_MAPPING = {
    "ternary": TernaryMask,
    "shearing": ShearingMask
}

@dataclass
class MaskConfig:
    mask_type: Literal["ternary", "shearing"]
    d_hidden: int
    d_inter: int
    n_heads: int
    n_layers: int
    d_hidden_target: int
    d_inter_target: int
    n_heads_target: int
    n_layers_target: int

class MaskGroup(nn.ModuleDict):
    def __init__(self, cfg: MaskConfig):
        super().__init__()

        self.n_layers = cfg.n_layers

        mask_class = MASK_CLASS_MAPPING[cfg.mask_type]

        self.masks_hidden = nn.ModuleList(
            [mask_class(1, cfg.d_hidden) for _ in range(self.n_layers)])
        self.masks_inter = nn.ModuleList(
            [mask_class(1, cfg.d_inter) for _ in range(self.n_layers)])
        self.masks_heads = nn.ModuleList(
            [mask_class(1, cfg.n_heads, 1, 1) for _ in range(self.n_layers)])
        self.masks_layer = nn.ModuleList(
            [mask_class(1) for _ in range(self.n_layers)])

        self.d_hidden_target = cfg.d_hidden_target
        self.d_inter_target = cfg.d_inter_target
        self.n_heads_target = cfg.n_heads_target
        self.n_layers_target = cfg.n_layers_target
    
    def compute_num_layers_constraint(self):
        lagmul_1 = self.masks_layer[0].lagmul_1
        lagmul_1 = 2 * lagmul_1.detach() - lagmul_1
        
        lagmul_2 = self.masks_layer[0].lagmul_2
        lagmul_2 = 2 * lagmul_2.detach() - lagmul_2
        
        c = sum(mask.n for mask in self.masks_layer) - self.n_layers_target
        
        return lagmul_1 * c + lagmul_2 * c.pow(2)
        
    def compute_constraint(self):
        p1 = sum(
            mask.compute_constraint(self.d_hidden_target)
            for mask in self.masks_hidden
        )
        p2 = sum(
            mask.compute_constraint(self.d_inter_target)
            for mask in self.masks_inter
        )
        p3 = sum(
            mask.compute_constraint(self.n_heads_target)
            for mask in self.masks_heads
        )
        
        p4 = self.compute_num_layers_constraint()

        return p1 + p2 + p3 + p4

**Comment by jeafatkamano** [comment]
Hi !
